[Rdy] Quick fix for garlic cloves decay time

nexusmrsep opened the pull request:
@FulcrumA noticed in #24145 that garlic cloves do not rot as garlic bulbs do, and this PR fixes that.

[edit:] Also see discussion below and subsequent commits.

Kelenius commented:
I think you need to add proper garlic seeds before making the only thing that can be planted spoil in an year, or it becomes hard to preserve the seeds for planting.

nexusmrsep commented:
Good point. What would be better? Replacing garlic cloves entirely with garlic seeds, or introducing separate garlic seeds, and transforming garlic cloves to a non-seed item (might require some recipes reworks)?

Kelenius commented:
As far as I understand, cloves and seeds should be distinct items, both of them should be plantable, and the resulting plant should produce garlic **bulbs** and garlic seeds. I'm not sure how easily doable that is with the current code.

nexusmrsep commented:
With a bit effort I think its doable, but might get complicated enough to split into 2 PR. I'll look into that.

TechyBen commented:
Seeds and bulbs have different viability and growth speeds/fruition speeds. AFAIK.

Amariithynar commented:
Yeah, seeds and bulbs are different because the bulbs are basically just their root system, just dug up. This should be standard for every root vegetable, tbqh. (the seeds and root being separate and the root not being craftable into the other).

Also, seeds DO have an expiry date, though they can be preserved indefinitely in some very niche extreme cases; there's one case of a 2000 year old date palm seed growing, for example. In any normal Survivor situation, it's just not feasible to replicate. That said, their rot timers would be measured in like, multiples of years, so it's not like there's any rush to add them for them.

FulcrumA commented:
>That said, their rot timers would be measured in like, multiples of years, so it's not like there's any rush to add them for them.

I suspect rarely any game will be lasting that long, with single-digit exceptions, it'd be just more things to track needlessly. I prefer nexusmrsep's cut for spoilage in #23986 - if something can last two years without significant drop in quality/properties, I'd be fine with making it last indefinitely. What we need is just more data in the relevant ticket on how long what should last (in said #23986) and how it should be affected by different preservation methods, containers etc (in #24145), based on sources found and verified by a professional (hoping for your further and continuous input here, @Amariithynar).

I do agree that there should be bigger differentiation between seeds and bulbs/stems (also, related, I'd reckon wild vegetable stems and similar should also spoil but then I'd do without such generic item as "wild vegetables" and introduce actual wild vegetable-analogs). I believe there's even different growth rate between such. Recipes to acquire different "reproductibles" should still be all in Food > Seeds category though.

nexusmrsep commented:
Ok, as per request, I've separated garlic seeds from garlic cloves, leaving the cloves as a separate comestible. I've also added "Split" action to garlic bulb so you can split it to 6 garlic cloves, and removed garlic bulb from recipes.

For a dependency reason yet unknown to me I could not remove recipe for garlic seeds (made from garlic bulb) without causing runtime error in `recipe_dictionary.cpp`. If i find this reason I might remove it, but for now it stays.

nexusmrsep commented:
> Onions can have a half inch of the bulb cut off where the root is and that planted, and they'll grow. 

This not full reproductory cycle. There is no "onion bulb to onion bulb to onion bulb" way without the seeds.
![image](https://user-images.githubusercontent.com/39925111/42418831-27342a64-82a9-11e8-82f2-552e010a7696.png)

nexusmrsep commented:
Also:
> If recipes are going to use a bulb's worth of cloves at all times anyways, is there a point for bulbs to continue to be rendered into cloves separately, now that seeds are separate from the bulb? Is the amount per being reduced (I usually use no more than half a bulb in any given recipe, for example, and then only if I want it to be really garlicky)?

Many recipes use just few cloves, not whole bulb worth of cloves. I encourage you to read the code.

 > Or are cloves still plantable individually, which they can (one source among many: http://www.southernexposure.com/blog/2010/11/how-to-grow-garlic-from-a-clove/ )?

Nope, I went for them being the ingredient, while the bulb is no longer the ingredient, and seeds are seeds. Not everything has to be simulated here. So: you plant the seed, get bulb(s). You can convert the bulb to seeds via recipe, or you just split the bulb to cloves and use them for cooking. This I think is as close to RL as we want to get without an overkill.

Kelenius commented:
>You mean that dust left from about 1/2 of cloves from my year old garlic bulbs that I held in my kitchen

I thought we're supposed to use sources instead of personal anecdotes.

nexusmrsep commented:
> I thought we're supposed to use sources instead of personal anecdotes.

Enlighten me then, and give me the source for garlic being "basically last indefinitely when properly stored". I'll give you a handicap: let's not go past say 50 years.

But seriously: No offence, but a majority of what I hear here and in "future of food preservation" are personal anecdotes. As long as discussion goes, I'm ok with that, as there is value to that, but when it comes to values i'm not so liberal at all, as you clearly see we don't agree on many matters, or have our own opinion on how things should be in game, so an external objective source is far better then opinions. I'll try to restrain myself from feeding the flame from now, because this is a development discussion not a forum, but will never support ideas that have nothing to do with life experience.

TechyBen commented:
(A lot of Google searching to double check my observations on the allotment).

Nope, both are viable. Planting cloves (individual parts of the bulbs), or planting seeds (bulbils).
See: https://en.wikipedia.org/wiki/Garlic#Cultivation

It might be beyond this PR, or out of scope of the game simulation (where do we stop on granularity?). However, it is not because seeds or bulbs don't exist as viable methods. In fact a lot of the plants in game revert to "seed" when bulb/root/splitting would be the only viable method to most untrained people.

nexusmrsep commented:
Seems this time you did my homework @TechyBen Thanks!
So: do we keep things as proposed or go for both methods?
I don't see any reason why not to, as we already have both and I can revert the part where i cut seed info from cloves.

TechyBen commented:
@nexusmrsep Sorry, my bad! I missed your link, it mentions all of that already! But I had not realised which direction you were taking this.

Cultivated garlic rarely has seed, as your link example shows, so in game, would it be "wild garlic"? No change needed if labeled to the player as "wild" and thus seeding variety. :+1:

nexusmrsep commented:
That was before I knew that planted cloves can yield a full bulb...
Third option is to revert changes and leave the cloves as seeds and ingredients, as they were.
I'm confused now.

TechyBen commented:
From a gameplay/GUI perspective. Having cloves appear in food and seed does not seem possible in sort lists? So we would need to pick one for sorting? Sunflower seeds currently are in the seed option, even though they can also be eaten.

Is there any other food we treat as both a seed and a food? Taking Rhubarb into account... I don't think we have it correct (rhubarb -> plantable stems), it would be the other way around (plantable root + stem cut down -> eatable stems).

So, would you prefer to put the types back. Keep the new rot mechanics... and put the seed/plant/stem/root mechanics on a seperate PR? Then all root/seed/stem style foods/plants can be worked on in one go?

Amariithynar commented:
>This not full reproductory cycle. There is no "onion bulb to onion bulb to onion bulb" way without the seeds.

You cut off the bulb about a 1/4th inch below the root and replant the rest of the plant, as I said. This is fairly common knowledge, with plenty of available references out there as well. I mean, here's a video showing a woman getting two plants from a single onion bulb. https://www.youtube.com/watch?v=Et4lGgjUMfY 

>Many recipes use just few cloves, not whole bulb worth of cloves. I encourage you to read the code.

When I hit / to filter my recipes at Cooking 11, and type c:garlic clove, I get recipes that use 6 or 12 cloves, aka 1 or 2 bulbs. *shrug* If there's more recipes, maybe I'm missing some books? but that's why I asked. You can take your attitude here and shove it where the sun doesn't shine, though.

>You mean that dust left from about 1/2 of cloves from my year old garlic bulbs that I held in my kitchen was not dried mold and just "chemically flavored and generally unpalatable, but it's still edible, technically."? Laughable. There is no "basically last indefinitely when properly stored", its a wishful thinking. You preach for the immortal life of food when experience tells otherwise. Cloves dry out with time, yes, lose their properties, yes, but then they either dry out to inedibility or develop mold that leaves nothing behind. There is no "indefinite" in this, just "long".

Stored in a cool dry place- and I do mean DRY, not "in a cupboard somewhere in an apartment in the middle of a city"- they desiccate and shrivel up into a hard rock-like node that can be rehydrated and used again. I've personally done this with 3 year old cloves a few years back that were stored in a heat-sealed bag after dehydrating somewhat in my fridge when I didn't use all of a bulb's cloves and stuff was put in front of it by my roommate (I wanted to experiment when I found them and see how long it would last; used them after three years just to try it out). It's far more likely that the conditions of storage aren't going to be conducive to long-term storage, but it can be done.

>No offence, but a majority of what I hear here and in "future of food preservation" are personal anecdotes...

If I've given a personal anecdote on something, it's because it's just that, a personal experience I've had with food while working in kitchens, residential, industrial (I guess it'd be more accurate to refer to this as "institutional"? w/e) and commercial, most of my life. These aren't things I've learnt from just sitting around reading books, or reading an article about it, but by actually having to deal with it, or talking with the farmers who sold us produce and finding out interesting things about them. Most of the "book learning" part of it is regarding the FDA-and-similar bodies related information, such as how they overly emphasize food safety and customer experience over actual food edibility. It's not like food waste stats are hidden and not trumpeted about all the time, especially with the wasteful bleaching process that supermarkets are forced to do by law to prevent dumpster diving for what would otherwise be still edible. That FDA-and-similar info came from the course you have to take to get your initial certification, and also to renew your certification, where they very specifically lay it all out for you in proprietary books that you have to pay for, in person, to access (or at least our courses have), provided specifically for the course. If you have a problem with that, tough.

nexusmrsep commented:
@Amariithynar I've no intention to disrespect you or your experience. I apologize if you feel I did.
But my experience is sometimes different form your, so don't blame me for clashing what I know with what you know. Also as I said - anecdotes are good for this discussion, as they often clear many misconceptions, even on my side. I don't have the patent for knowledge. But sources are invaluable if there are different experiences involved or when a decision must be made under uncertainty.

Also, if you can help filter the no-freeze list of foods in #24228 and/or give good sources for it I'd be more then happy.

As for onions - i have following questions (bit off topic but still valuable knowledge for our "seed" problem):

[Wikipedia ](https://en.wikipedia.org/wiki/Onion)says:
> Onions may be grown from seed or from sets. Onion seeds are short-lived and fresh seeds germinate better. The seeds are sown thinly in shallow drills, thinning the plants in stages. In suitable climates, certain cultivars can be sown in late summer and autumn to overwinter in the ground and produce early crops the following year. Onion sets are produced by sowing seed thickly in early summer in poor soil and the small bulbs produced are harvested in the autumn. These bulbs are planted the following spring and grow into mature bulbs later in the year. Certain cultivars are used for this purpose and these may not have such good storage characteristics as those grown directly from seed.

It seems different form garlic, where it directly says it grows both from cloves and seeds.
If so, then for onions: Can this regrowth process go forever? Can you get multiple bulbs or just a regrow one bulb? Is it viable process for in-game purpose?

nexusmrsep commented:
Also @Amariithynar you think I should revert the changes here, and go back to only planting cloves?

Amariithynar commented:
>I've no intention to disrespect you or your experience. I apologize if you feel I did.
But my experience is sometimes different form your, so don't blame me for clashing what I know with what you know. Also as I said - anecdotes are good for this discussion, as they often clear many misconceptions, even on my side. I don't have the patent for knowledge. But sources are invaluable if there are different experiences involved or when a decision must be made under uncertainty.

When your experience is different, that's fine. I've had garlic rot on me too- note that I've always said that they can keep *when properly stored and undamaged*, not just sitting out on the floor, or without being climate-controlled (or happening to be in a climate-controlled location). I've never invalidated your own experiences because I know that there's a wide range of possibilities- only argued against statements made wherein my own experiences or knowledgebase does not agree. I've never taken umbrage at that. I even explicitly stated how fine I was with utilizing the FDA-and-similar-federal-bodies' "best before" data for spoilage even though it directly contradicted what I had been taught and knew, so long as the choice made was an informed one about how it's about how business operates at a commercial level and not a strict, "past this point it's completely inedible" sort of statement by them. What I took umbrage with was your dismissive attitude towards me and most of my knowledge being directly earned rather than being scientific papers written up somewhere that I can easily source, nothing more. Either way, apology accepted, thank you.

>Also, if you can help filter the no-freeze list of foods in #24228 and/or give good sources for it I'd be more then happy.

I'll take a look at the list of no-freeze items, sure. 

>As for onions - i have following questions...
Commercially available sets are effectively those new growth onions that grow from within the bulb; onions that have been grown once and are on their second growth. In the video I linked, those would be the two green stalks growing out of the bulb that she cuts the flesh of the bulb away from and carefully separates to plant, just like how each individual clove in a garlic bulb can sprout and grow a new plant to increase the total number of plants and thus the potential for success of that particular evolutionary tree in propagating. In fact, onions are a biennial, where the bulb primarily grows in the first year after planting, goes dormant until the next growth cycle, and then flowers (and disperses seeds) in the second. By this nature, their bulbs won't grow to be as large as they were the first year, but they're still quite edible, and the onion greens are also quite edible, too. An onion grown to seed has the (aboveground) plant die off at the end of the cycle, but if you harvest the bulb and follow the same process, then it will keep regrowing indefinitely (effectively being converted to a perennial due to human interference), so long as it has the nutrients to do so, and doesn't get so damaged that it can't recover before rotting away (there will always be some rotten bits around the bulb when the rooting method is done, as that's what the layers are for, and why they produce a chemical to make things that bite into them tear up as a deterrent). Personally I think that root vegetable greens are vastly underutilized in-game, even if just to make something like a "farmer's salad" that doesn't taste very good, but is nutritious and provides a variety of vitamins. Many plants can be regrown like this, where either the core for new life of the plant isn't what is consumed or where you grow it from the root bottom that you'd cut off anyways because it isn't palatable; here's a short list of a few. https://www.diyncrafts.com/4732/repurpose/25-foods-can-re-grow-kitchen-scraps

>you think I should revert the changes here, and go back to only planting cloves?

I'd rather just see cloves being able to be planted as well OR being converted into "seeds" at a 1:1 rate (abstracting out that the cloves can also be planted to grow a new plant), and getting seeds when you harvest the plant, not getting seeds directly from the bulbs.

nexusmrsep commented:
Done & done. Quick test positive, you can plant both cloves and seeds.

geophree commented on data/json/items/comestibles/seed.json:
```diff
@@ -197,14 +197,30 @@
     "price": 50,
+    "name": "garlic seeds",
+    "name_plural": "garlic seeds",
+    "description": "Some onion seeds.",
```
This should probably read "Some garlic seeds."

geophree commented on data/json/items/comestibles/seed.json:
```diff
@@ -197,14 +197,30 @@
     "nutrition": 3,
-    "weight": 3,
+    "description": "Cloves of garlic.  Useful as a seasoning.",
+    "price": 50,
```
If we can still plant garlic seeds, should you leave in the "or for planting" part of the description?
